clarity/src/vm/tests/traits.rs

```diff
@@ -1281,3 +1286,345 @@ fn test_return_trait_with_contract_of(owned_env: &mut OwnedEnvironment) {
+                "wrapped-get-1",
+                &symbols_from_values(vec![target_contract]),
+                false
+            )
+            .unwrap(),
+            Value::okay(Value::UInt(1)).unwrap()
+        );
+    }
+}
+
+fn test_embedded_trait(owned_env: &mut OwnedEnvironment) {
+    let dispatching_contract = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-public (may-echo (opt (optional <trait-1>)))
+            (match opt
+                t (contract-call? t echo u42)
+                (err u1)
+            )
+        )";
+    let target_contract = "(define-public (echo (a uint)) (ok a))";
+
+    let p1 = execute("'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR");
+    let mut placeholder_context = ContractContext::new(
+        QualifiedContractIdentifier::transient(),
+        ClarityVersion::Clarity2,
+    );
+
+    {
+        let mut env = owned_env.get_exec_environment(None, None, &mut placeholder_context);
+        env.initialize_contract(
+            QualifiedContractIdentifier::local("dispatching-contract").unwrap(),
+            dispatching_contract,
+        )
+        .unwrap();
+
+        env.initialize_contract(
+            QualifiedContractIdentifier::local("target-contract").unwrap(),
+            target_contract,
+        )
+        .unwrap();
+    }
+
+    {
+        let target_contract = Value::from(PrincipalData::Contract(
+            QualifiedContractIdentifier::local("target-contract").unwrap(),
+        ));
+        let opt_target = Value::some(target_contract).unwrap();
+        let mut env = owned_env.get_exec_environment(
+            Some(p1.clone().expect_principal()),
+            None,
+            &mut placeholder_context,
+        );
+        assert_eq!(
+            env.execute_contract(
+                &QualifiedContractIdentifier::local("dispatching-contract").unwrap(),
+                "may-echo",
+                &symbols_from_values(vec![opt_target]),
+                false
+            )
+            .unwrap(),
+            Value::okay(Value::UInt(42)).unwrap()
+        );
+    }
+}
+
+fn test_pass_embedded_trait_to_subtrait(owned_env: &mut OwnedEnvironment) {
+    let dispatching_contract = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-12 (
+            (get-1 (uint) (response uint uint))
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-opt-get-1 (contract <trait-12>))
+            (wrapped-get-1 (some contract)))
+        (define-public (wrapped-get-1 (opt-contract (optional <trait-1>)))
+            (internal-get-1 opt-contract)
+        )
+        (define-public (internal-get-1 (opt-contract (optional <trait-12>)))
+            (match opt-contract
+                contract (contract-call? contract get-1 u1)
+                (err u1)
+            )
+        )";
```

Comment:
We're sure this also works for `(ok <trait-1>)` and `(err <trait-1>)` and `(list 1 <trait-1>)`?

Reply:
I can add tests for those too.

Reply:
Added tests and fixed an issue with lists in 748611aad.